Lib/idlelib/formatregion.py

```diff
@@ -0,0 +1,178 @@
+"""Format a selected region of text.
+
+Several formatting options are available:
+indent, deindent, comment, uncomment, tabify, and untabify.
+"""
+from tkinter.simpledialog import askinteger
+
+
+def classifyws(s, tabwidth):
+    """Classify leading whitespace in a string.
+
+    Count the number of actual spaces and tabs at the beginning of
+    the string.  Also count the tabwidth number of spaces that a
+    tab would expand to.
+
+    Return a tuple of (number of raw leading whitespace characters,
+        number of leading characters after expanding tabs).
+    """
+    raw = effective = 0
+    for ch in s:
+        if ch == ' ':
+            raw = raw + 1
+            effective = effective + 1
+        elif ch == '\t':
+            raw = raw + 1
+            effective = (effective // tabwidth + 1) * tabwidth
+        else:
+            break
+    return raw, effective
+
+
+class FormatRegion:
+    "Format selected text."
+
+    def __init__(self, editwin):
+        self.editwin = editwin
+
+    def get_region(self):
+        """Return line information about the selected text region.
+
+        If text is selected, the first and last indices will be
+        for the selection.  If there is no text selected, the
+        indices will be the current cursor location.
+
+        Return a tuple containing (first index, last index,
+            string representation of text, list of text lines).
+        """
+        text = self.editwin.text
+        first, last = self.editwin.get_selection_indices()
+        if first and last:
+            head = text.index(first + " linestart")
+            tail = text.index(last + "-1c lineend +1c")
+        else:
+            head = text.index("insert linestart")
+            tail = text.index("insert lineend +1c")
+        chars = text.get(head, tail)
+        lines = chars.split("\n")
+        return head, tail, chars, lines
+
+    def set_region(self, head, tail, chars, lines):
+        """Replace the text between the given indices.
+
+        Args:
+            head: Starting index of text to replace.
+            tail: Ending index of text to replace.
+            chars: Expected to be string of current text
+                between head and tail.
+            lines: List of new lines to insert between head
+                and tail.
+        """
+        text = self.editwin.text
+        newchars = "\n".join(lines)
+        if newchars == chars:
+            text.bell()
+            return
+        text.tag_remove("sel", "1.0", "end")
+        text.mark_set("insert", head)
+        text.undo_block_start()
+        text.delete(head, tail)
+        text.insert(head, newchars)
+        text.undo_block_stop()
+        text.tag_add("sel", head, "insert")
+
+    def indent_region_event(self, event=None):
```

Comment:
This and the other methods below have so much in common that they should probably be refactored.

Reply:
I agree, but I wanted to make this PR about removing them from editor.py, so I kept them as they were currently written.  Once the code has been split from editor.py and the tests are in place, then I'll work on refactoring.